docs/05-rfcs/2023-01-11-sdk-spec.md

```diff
@@ -0,0 +1,1070 @@
+
+> Reqtag: `sdk:on-event-name`
+<span id="sdk:on-event-name"/>
+
+For example, a `cloud.Bucket` can emit events whenever an object is uploaded, and these events can be listened to by calling `on_upload` on the bucket:
+
+```ts
+// wing
+let bucket = new cloud.Bucket();
+bucket.on_upload(inflight (file: str) => {
+  log.info("File uploaded: " + file);
+});
+```
+
+The argument to an `on_<event>` method is called an event handler.
+An event handler is any resource that implements the `EventHandler<T>` interface, which contains a single method named `handle` that accepts an event of type `T`:
+
+```ts
+// wing
+
+interface EventHandler<T> {
+  inflight handle(event: T);
+}
+```
+
+<!--
+We could also decide the should be named `handle_event` or `invoke` or `run` or `call` or something else.
+-->
+
+An inflight closure that accepts a single argument of type `T` is also considered an event handler, since the compiler will automatically convert it into a resource that implements the `EventHandler<T>` interface.
+For example, the definitions of `x1` and `x2` are equivalent:
+
+```ts
+// wing
+let x1 = inflight (event: MyEvent) => {
+  log.info("Event received: " + event);
+};
+
+resource X implements EventHandler<MyEvent> {
+  inflight handle(event: MyEvent) {
+    log.info("Event received: " + event);
+  }
+}
+let x2 = new X();
+```
+
+> Reqtag: `sdk:event-handler-conversion`
+<span id="sdk:event-handler-conversion"/>
+
+In the example with `bucket.on_upload` from earlier this section, when the method is called, cloud infrastructure is automatically added to invoke the event handler's `handle` method whenever an object is added to the bucket.
+
+Users can define their own event-based APIs by making use of the `Topic` resource.
+For example, suppose you want to extend `cloud.Counter` so that it triggers an event whenever a threshold is reached or exceeded:
+
+```ts
+// wing
+
+let counter = new MyCounter(threshold: 10);
+counter.on_threshold(inflight (event: ThresholdReachedEvent) => {
+  log.info("Threshold reached: " + event.value);
+});
+```
+
+You can implement this by defining a `CounterWithThreshold` resource with an `on_threshold` method, and use a `Topic` to publish events:
+
+```ts
+// wing
+
+struct MyCounterProps extends cloud.CounterProps {
+  threshold: num;
+}
+
+resource MyCounter extends cloud.Counter {
+  _threshold: num;
+  _topic: cloud.Topic;
+
+  init(props: CounterWithThresholdProps) {
+    super(props);
+    this._threshold = props.threshold;
+    this._topic = new cloud.Topic();
+  }
+
+  inflight inc(amount?: num) {
+    let new_value = super.inc(amount);
+    if new_value >= this._threshold {
+      this._topic.publish(ThresholdReachedEvent {
+        value: new_value,
+        time_reached: time.now(),
+      });
+    }
+    return value;
+  }
+
+  // or: on_threshold(handler: EventHandler<ThresholdReachedEvent>)
+  on_threshold(handler: inflight (event: ThresholdReachedEvent) => void): void {
```

Comment:
I think the second void can be inferred by the compiler (based on type inference on the body of the function) but not necessarily the first one.

BTW, maybe it should be `inflight (event: ThresholdReachedEvent): void` (no arrow) since it's just a type annotation... but I suppose we could define it to be either way in our grammar.

Reply:
I removed the second `void` but left it in most other places for emphasis- let me know if you think it should be otherwise.